### PR TITLE
Scheduler.

### PR DESCRIPTION
**Pull request by floitschG**
I'm not sure about the names.
There is already a `scheduler` in flutter/animation, so I changed this one to `tasks`. This makes a nice `tasks.schedule(...)`.
Also, having a typedef `Task` for a simple thunk feels slightly overkill.

I think I have seen `kFoo` for constants. That's not standard Dart, and if I'm wrong (i.e. Flutter is not using that convention), I'm happy to change it.

There are still lots of opportunities to make this better. The most important one is probably to take touch-events into account. I prefer adding this in a separate CL.

### CODE REVIEW

**Comment by Hixie** [comment]
Make sure to read 
* https://github.com/flutter/engine/blob/master/sky/specs/style-guide.md
* https://github.com/flutter/engine/blob/master/sky/specs/design.md
(Note in particular that the Flutter style guide differs in some significant ways from the Dart style guide.)

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 1]
```diff
@@ -0,0 +1,149 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:collection/priority_queue.dart';
+import 'package:flutter/animation.dart' as animation show scheduler;
+
+typedef Task();
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: needs explicit return type.
If this is void, then use `VoidCallback` from `dart:ui`.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 2]
```diff
@@ -0,0 +1,149 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:collection/priority_queue.dart';
+import 'package:flutter/animation.dart' as animation show scheduler;
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Where is this intended to fit in the layering?:
https://docs.google.com/presentation/d/1cw7A4HbvM_Abv320rVgPVGiUP2msVs7tfGbkgdrTy0I/edit#slide=id.gbb3c3233b_0_162

File: packages/flutter/pubspec.yaml [hunk 3]
```diff
@@ -4,6 +4,7 @@ author: Flutter Authors <flutter-dev@googlegroups.com>
 description: A framework for writing Flutter applications
 homepage: http://flutter.io
 dependencies:
+  collection: '>=1.1.3 <2.0.0'
```
Comment (Hixie) [inline packages/flutter/pubspec.yaml]: cc @abarth, who may have opinions on adding new dependencies

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+typedef Task();
+
+/// An entry in the scheduler's priority queue.
+///
+/// Combines the task and its priority.
+class _SchedulerEntry {
+  final Task task;
+  final int priority;
+
+  _SchedulerEntry(this.task, this.priority);
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: constructor first, then fields.
If all the fields are final, then make the constructor const.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  /// It is still possible to have priorities that are offset by more than this
+  /// amount by repeatedly taking relative offsets, but that's generally
+  /// discouraged.
+  static const kMaxOffset = 10000;
+
+  final int _value;
+
+  int get value => _value;
+
+  const Priority._(this._value);
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: constructor should go first

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+/// Combines the task and its priority.
+class _SchedulerEntry {
+  final Task task;
+  final int priority;
+
+  _SchedulerEntry(this.task, this.priority);
+}
+
+class Priority {
+  static const kIdle = const Priority._(0);
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: these need explicit types

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  /// Relative priorities are clamped by this offset.
+  ///
+  /// It is still possible to have priorities that are offset by more than this
+  /// amount by repeatedly taking relative offsets, but that's generally
+  /// discouraged.
+  static const kMaxOffset = 10000;
+
+  final int _value;
+
+  int get value => _value;
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: put the getter just above the field, so that when we /// dartdoc the getter, the comment is in a block just above the getter and the field (otherwise the dartdocs comments end up splitting the field and the getter)

File: packages/flutter/pubspec.yaml (hunk 3 above)
Comment (abarth) [inline packages/flutter/pubspec.yaml]: collection doesn't have any transitive dependencies, so its probably ok.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+  const Priority._(this._value);
+
+  /// Returns a priority relative to this priority.
+  ///
+  /// A positive [offset] indicates a higher priority.
+  ///
+  /// The parameter [offset] is clamped to +/-[kMaxOffset].
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: What is the purpose of kMaxOffset? Is it to reduce the likelihood of priority inflation?

Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Yes.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+/// Combines the task's priority with remaining time in a frame to decide when
+/// the task should be run.
+///
+/// Tasks always run in the idle time after a frame has been committed.
+class TaskScheduler {
+  final PriorityQueue _queue = new HeapPriorityQueue<_SchedulerEntry>(
+    (_SchedulerEntry e1, _SchedulerEntry e2) {
+      // Note that we inverse the priority.
+      return -e1.priority.compareTo(e2.priority);
+    });
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: final ) should be on its own line

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+  SchedulingStrategy schedulingStrategy = new SchedulingStrategy();
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: If this is meant to be a hook for people to change the strategy, then I would consider having the interface named separately than the default implementation. e.g. have a `DefaultSchedulingStrategy` that implements `SchedulingStrategy`.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+      // to run in the idle-phase of an animation. We therefore request to be
+      // woken up as soon as possible.
+      _wakeNow();
+    });
+  }
+}
+
+final TaskScheduler tasks = new TaskScheduler._();
+
+class SchedulingStrategy {
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Why a class rather than a callback? It's only got one entrypoint.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+  SchedulingStrategy schedulingStrategy = new SchedulingStrategy();
+
+  /// Wether this scheduler already requested to be woken up as soon as
+  /// possible.
+  bool _wakingNow = false;
+  /// Wether this scheduler already requested to be woken up in the next frame.
+  bool _wakingNextFrame = false;
+
+  TaskScheduler._();
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: constructor goes first

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+  /// Schedules the given [task] with the given [priority].
+  // TODO(floitsch): provide some means to indicate how long things are going
+  // to take?
+  // TODO(floitsch): guidance on how to increase priority over time?
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: I would either remove these todos (if we have no use case, then there's no need to add it; when we have a use case, we can add it even without the todo), or rephrase it as "other features that could be added include...". We try to avoid checking in questions.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+  void schedule(Task task, Priority priority) {
+    bool isFirstTask = _queue.isEmpty;
+    _queue.add(new _SchedulerEntry(task, priority._value));
+    if (isFirstTask) _wakeNow();
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: newline after expression before `_wakeNow();`

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  /// Invoked by the system when there is time to run tasks.
+  void tick() {
+    if (_queue.isEmpty) return;
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: newline before return

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+    _SchedulerEntry entry = _queue.first;
+    if (schedulingStrategy.shouldRunTaskWithPriority(entry.priority)) {
+      try {
+        (_queue.removeFirst().task)();
+      } finally {
+        if (_queue.isNotEmpty) {
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: no braces around one-line blocks

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+      }
+    } else {
+      _wakeNextFrame();
+    }
+  }
+
+  /// Tells the system that the scheduler is awake and should be called as
+  /// soon a there is time.
+  void _wakeNow() {
+    if (_wakingNow) return;
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: newline before return

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+    Timer.run(() {
+      _wakingNow = false;
+      tick();
+    });
+  }
+
+  /// Tells the system that the scheduler needs to run again (ideally next
+  /// frame).
+  void _wakeNextFrame() {
+    if (_wakingNextFrame) return;
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: newline before return

File: packages/flutter/lib/task.dart [hunk 19]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
Comment (Hixie) [inline packages/flutter/lib/task.dart]: As per our discussion, I think this should probably be "scheduler" and should import the scheduler from the animation layer.

You may wish to do this as a separate CL (maybe before this one, maybe after, whichever is easiest) since that CL will involve a lot of changes to other files.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 10 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: That's what I had :)
I changed it, because I wasn't sure of your style. Will change it back.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 11 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Initially it had another callback: `shouldContinueRunningTaskWithPriority`

I'm not 100% if that's still necessary. I would like to provide a kLowPriority which can be interrupted at microtasks (when the user invokes `scheduleMicrotask` or `await`). In this case it could make sense to use a separate function.

I think I will go with a simple function callback, though.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+final TaskScheduler tasks = new TaskScheduler._();
+
+class SchedulingStrategy {
+  // TODO(floitsch): for now we only expose the priority. It might be
+  // interesting to provide more info (like, how long the task ran the last
+  // time).
+  bool shouldRunTaskWithPriority(int priority) {
+    if (animation.scheduler.transientCallbackCount > 0) {
```
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: So here's a case we should consider.

Suppose you have an app that is trying to do a network load.
It queues a bunch of tasks for this, all at a priority lower than kAnimation.
It then displays an indeterminate progress spinner.

The spinner will spin every frame, effectively starving the network tasks, despite the fact that the system isn't at all busy (moving the spinner leaves many free milliseconds per frame).

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 6 above)
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: You probably don't need the "k"s here. "Priority.idle", "Priority.animation", "Priority.touch", etc, is probably fine.

**Comment by Hixie** [comment]
This looks great.

One problem we will have to figure out is how to integrate this with Timers and microtasks. Can we add a priority field to the Timer constructor somehow, for instance, that would integrate with this? It would be unfortunate to have this mechanism but then have everyone just use "Timer" APIs everywhere and preempt all this logic. (imagine devs talking: "My tasks aren't running." "Oh use `new Timer()`, it'll run faster".)

**Comment by floitschG** [comment]
I don't think we should change the `Timer` and `scheduleMicrotask` signatures by adding a new argument. Even optional, it feels very brittle to me.

The embedder is responsible for setting up the Timer and microtask callbacks for the system. If could provide an API allowing Dart code to intercept these calls.
For Timer that would work relatively easily, for microtasks it's more complicated, since the microtask callback is only happening when the internal queue is drained (and an error has been encountered).

I'm not really sure how I feel about this, though.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 2 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: I will merge with the animation scheduler, in which case it will be below animation (but above services).

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 6 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Just found the relevant section in the style guide. Since it's not a string or double, I will change them.

Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Yeah that part of the style is a bit more ad hoc than designed.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 1 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done. (VoidCallback)

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 4 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.
Note: removing `const` from a constructor is a breaking change. You should only add it, if you are willing to commit to it. (In this case it's an internal class, so it's fine).

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 5 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: The style-guide doesn't say if the constructor should be before or after "static".
I found other code in flutter where the statics are first, so I moved the constructor between the statics and the fields.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 6 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done (explicit types).

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 7 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 9 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 12 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 13 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done. (removed).

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 14 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 15 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 16 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 17 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 18 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.

File: packages/flutter/lib/task.dart (hunk 19 above)
Comment (floitschG) [inline packages/flutter/lib/task.dart]: Will do in a separate CL.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 20 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: :(

As discussed in person: this will require serious rethinking.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+  static const Priority idle = const Priority._(0);
+  static const Priority animation = const Priority._(100000);
+  static const Priority touch = const Priority._(200000);
+
+  /// Relative priorities are clamped by this offset.
+  ///
+  /// It is still possible to have priorities that are offset by more than this
+  /// amount by repeatedly taking relative offsets, but that's generally
+  /// discouraged.
+  static const kMaxOffset = 10000;
```
Comment (abarth) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Please add the missing type.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+  bool _wakingNow = false;
+
+  /// Wether this scheduler already requested to be woken up in the next frame.
+  bool _wakingNextFrame = false;
+
+  /// Schedules the given [task] with the given [priority].
+  void schedule(VoidCallback task, Priority priority) {
+    bool isFirstTask = _queue.isEmpty;
+    _queue.add(new _SchedulerEntry(task, priority._value));
+    if (isFirstTask) _wakeNow();
```
Comment (abarth) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: new line for if-body pls.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 23]
```diff
@@ -0,0 +1,151 @@
...
+  void tick() {
+    if (_queue.isEmpty)
+      return;
+    _SchedulerEntry entry = _queue.first;
+    if (schedulingStrategy.shouldRunTaskWithPriority(entry.priority)) {
+      try {
+        (_queue.removeFirst().task)();
+      } finally {
+        if (_queue.isNotEmpty)
+          _wakeNow();
```
Comment (abarth) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Do we need to go through the event loop for each task?  I guess the concern is that there might be higher priority work in the event queue?

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 21 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 22 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: done.

File: packages/flutter/lib/src/scheduler/scheduler.dart [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+  void _wakeNextFrame() {
+    if (_wakingNextFrame)
+      return;
+    _wakingNextFrame = true;
+    animation.scheduler.requestAnimationFrame((_) {
+      _wakingNextFrame = false;
+      // RequestAnimationFrame calls back at the beginning of a frame. We want
+      // to run in the idle-phase of an animation. We therefore request to be
+      // woken up as soon as possible.
+      _wakeNow();
```
Comment (abarth) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Why not schedule a post-frame callback?

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 23 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Going through the event-loop also allows the system to trigger a `beginFrame`.
We could avoid the event-loop by adding a way to see how much time there was left before the next `beginFrame`.
And yes: I would also like to give other event-loop operations the opportunity to run.

**Comment by abarth** [comment]
This looks like a great start.  It's probably worth integrating with the Mojo messaging system so that incoming mojo messages are prioritized.  That will let you, for example, kick off network requests and then have the replies prioritized.

Another direction is to teach the system how to schedule work during the gaps between frames.  We should look at some timeline traces together to get a sense for how that might work.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 24 above)
Comment (floitschG) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: Didn't know about it when I wrote the code.

I could switch, but there is the concern that we would run before other postAnimationFrame callbacks.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 23 above)
Comment (abarth) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: That makes sense.

Another thing we could consider is spinning the event loop inside Dart.  All the incoming signals to this thread are mojo handles and we have the MojoWaitMany primitive available in Dart.  IMHO, we should see how far we can get spinning the event loop in C++ though.

**Comment by floitschG** [comment]
Would be happy to discuss this tomorrow.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 24 above)
Comment (abarth) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: We can also elaborate the other scheduler.dart.  My point is just that going through the event loop here isn't what you want.  You want to run at defined time at the start of the idle period.  What you've got here is fine for the first iteration, but we'll eventually want to merge the two scheduling systems, and when we do, we'll probably change the timing of this callback.

**Comment by abarth** [comment]
LGTM

Happy to talk more tomorrow.  We should probably land this patch and iterate.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 4 above)
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: It's ok, we're a long way from committing to anything. It's easier to remove const from a class than add it (assuming you're willing to change the call sites, which for now we are), so we are tending towards const for now.

File: packages/flutter/lib/src/scheduler/scheduler.dart (hunk 5 above)
Comment (Hixie) [inline packages/flutter/lib/src/scheduler/scheduler.dart]: When in doubt, do whatever makes the code most readable.
